fix compare of foreign keys for autoupdate

darknos opened the pull request:
### Description
in case the name of a model is different from correspondent mysql table then foreign keys will be dropped and re-created again on autoupdate. this PR fix the comparison and rid out unnecessary drop/create on autoupdate

There was lost //TODO comment in a source code, so I just did a "todo".

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

darknos commented:
do you need any unit test for this?

darknos commented:
I already found three bugs for foreign key autoupdate so I feel I need more unit tests for it.

P.S. in our current project we allow user to change structure of Model (and correspondent mysql table) from UI so I should test all the ways.

dhmlau commented:
@darknos , the commit linter failed because the commit message is too long. 

```
**  Linting commit logs
**
**  2 problems found:
**    1c27e03 - if more the one foreign key added "ADD CONSTRAINT" : First line should be 50 characters or less (saw 121)
**    38707d0 - fix compare of foreign keys for autoupdate - use my: First line should be 50 characters or less (saw 100)
**
```

darknos commented:
Could you give me the linter config, because current .eslintrc ignores comments
```javascript
{
   "extends": "loopback",
   "rules": {
     "max-len": ["error", 120, 4, {
       "ignoreComments": true,
       ...
```

ssh24 commented on lib/migration.js:
```diff
@@ -423,7 +432,7 @@ function mixinMigration(MySQL, mysql) {
     }
     if (fkSQL && fkSQL.length) {
-      self.applySqlChanges(model, fkSQL, function(err, result) {
+      self.applySqlChanges(model, [fkSQL.join(', ')], function(err, result) {
```
Can we do something like `[fkSQL.toString()]` instead? I assume you are to concatenate all the fkSQL's and join them with `,`. The `toString()` function on an array does that same thing in a more elegant way.

darknos replied:
'elegant' is not equal to 'obvious', but I can change if you want.

darknos replied:
in the same file:
```
columns = indexedColumns.join(',');
```
```
sql.push('PRIMARY KEY(' + pks.join(',') + ')');
```

darknos replied:
@ssh24 please confirm replace `join` with `toString`.

btw, it should be more obvious to do
```javascript
self.execute(self.getAlterStatement(model, fkSQL), function(err, result) {
```
what do you think?

ssh24 replied:
I agree that `self.execute(self.getAlterStatement(model, fkSQL), function(err, result) {` would do what we were trying to. Other connectors use `applySqlChanges ` function and we should keep them consistent.

darknos replied:
ok. so `toString` or not `toString`? (c) Hamlet

ssh24 replied:
@darknos you just changed it as I was writing the comment. `toString()` it is :)

ssh24 commented on lib/migration.js:
```diff
@@ -449,7 +458,7 @@ function mixinMigration(MySQL, mysql) {
           var fkEntity = self.getModelDefinition(newFk.entity);
           var fkRefKey = expectedColNameForModel(newFk.entityKey, fkEntity);
-          var fkRefTable = newFk.entity.name; //TODO check for mysql name
+          var fkRefTable = self.table(newFk.entity);
```
Shouldn't this be `self.table(newFk.entity.name)`?

darknos replied:
@ssh24 
my foreign keys defined like:
```
            "FK_person_belongsTo_person_work": {
                "name": "FK_person_belongsTo_person_work",
                "entity": "PersonWork",
                "foreignKey": "personWorkId",
                "entityKey": "id"
            },
```

is this wrong way?

darknos replied:
I found above in the code
```
var fkEntityName = (typeof fk.entity === 'object') ? fk.entity.name : fk.entity;
```

I guess this is a correct way

ssh24 commented:
@slnode ok to test

darknos commented:
I see some tests failed. I need a time to find what happened.

darknos commented:
I found a reason. not my piece of code was wrong. please test now

ssh24 commented:
@slnode test please

dhmlau commented:
@darknos, there are 2 linting errors in migration.js.  Could you please take a look?
```
78:1   error  Line 78 exceeds the maximum line length of 120  max-len
  477:37  error  Missing space before opening brace              space-before-blocks
```

ssh24 commented:
@darknos Could you add some test cases?

darknos commented:
just released the service referenced to the fixed code :) , and now I'm working on tests.

darknos commented:
@ssh24 I added tests

darknos commented:
@ssh24 I added tests and rebased PR with the latest changes. I laready have to use this branch in production code because we need a lot of autoupdate calls for dynamic models.

please review and merge it.

ssh24 commented on test/mysql.autoupdate.test.js:
```diff
@@ -427,42 +424,48 @@ describe('MySQL connector', function() {
+                  assert(updatedForeignKeys);
+                  assert(updatedForeignKeys.length.should.be.equal(1));
+                  assert.equal(updatedForeignKeys[0].REFERENCED_TABLE_NAME, 'customer_test2');
+                  //assert.equal(updatedForeignKeys[0].COLUMN_NAME, 'customer_id');
```
Which is this commented out?

ssh24 replied:
Why is this assertion commented out?

darknos replied:
hmm... I'm not sure. probably accidentally  I need to check it.

darknos replied:
I uncommented it.

ssh24 commented:
@slnode test please

ssh24 commented:
@slnode test please

darknos commented:
@slnode test please

:)

ssh24 commented:
@slnode test please

ssh24 commented:
@slnode test please

ssh24 commented:
@darknos Thank you for the fix. PR is ready to land.
